### PR TITLE
Set proxy port in policy map

### PR DESCRIPTION
**Pull request by rlenglet**
Set the proxy port in policy map entries.
Reorganize `regenerateBPF` to keep ensuring that proxy redirects are always realized when redirecting traffic to them.

Output proxy port in `cilium bpf policy get` CLI, e.g.:
```
$ sudo cilium bpf policy get 29094
DIRECTION   LABELS (source:key[=value])   PORT/PROTO   PROXY PORT   BYTES   PACKETS   
Egress      container:id.client           ANY          NONE         0       0         
Egress      reserved:init                 ANY          NONE         0       0         
Egress      container:id.server           ANY          NONE         0       0         
            container:id.service1                                                          
Egress      reserved:world                ANY          NONE         0       0         
Ingress     reserved:host                 80/TCP       11555        0       0         
Ingress     container:id.client           80/TCP       11555        0       0         
Egress      reserved:host                 ANY          NONE         0       0         
Ingress     reserved:host                 ANY          NONE         0       0         
Egress      reserved:cluster              ANY          NONE         0       0         
Egress      reserved:health               ANY          NONE         0       0         
```

Remove the generation and lookup of the `CFG_L3L4_INGRESS` / `CFG_L3L4_EGRESS` arrays in BPF.
Only define `IPCACHE4_PREFIXES` / `IPCACHE6_PREFIXES` macros in BPF if LPM maps are not supported.
Therefore, on Linux kernels versions where LPM maps are supported (≥4.11), a policy regeneration shouldn't require a BPF regeneration anymore.

Fixes: https://github.com/cilium/cilium/issues/4711
Signed-off-by: Romain Lenglet <romain@covalent.io>

---
This change is [Reviewable](https://reviewable.io/reviews/cilium/cilium/4724)

### CODE REVIEW

**Comment by rlenglet** [comment]
test-me-please

**Comment by rlenglet** [comment]
I tested manually that `tests/10-proxy.sh` passes.

**Comment by rlenglet** [comment]
test-me-please

File: cilium/cmd/helpers.go [hunk 1]
```diff
@@ -269,7 +269,9 @@ func updatePolicyKey(cmd *cobra.Command, args []string, add bool) {
 		u8p := u8proto.U8proto(proto)
 		entry := fmt.Sprintf("%d %d/%s", label, port, u8p.String())
 		if add == true {
-			if err := policyMap.Allow(label, port, u8p, parsedTd); err != nil {
+			// TODO: Support passing a non-zero proxy port.
```
Comment (ianvernon) [inline cilium/cmd/helpers.go]: This is still a TODO, but should we ever do this? Wouldn't this require bootstrapping the proxy and having the proxy configured to accept traffic on that port?

**Comment by tgraf** [comment]
@rlenglet Let us know when this is ready for review

**Comment by rlenglet** [comment]
Ready for review.

**Review by tgraf — APPROVED** [review]
This is a great change as it requires less recompilation as well!

Looks great overall. In case you need to respin, consider using a struct for the value of `desiredMapState` for readability. It's not clear that the map value is the redirect port at first glance.

**Comment by rlenglet** [comment]
> In case you need to respin, consider using a struct for the value of desiredMapState for readability. It's not clear that the map value is the redirect port at first glance.

I'll do that. I need to fix the CLI anyway. We're currently not dumping the proxy ports from policy maps.

**Review by ianvernon — APPROVED** [review]

**Comment by ianvernon** [comment]
> This is a great change as it requires less recompilation as well!

@tgraf it looks like now that the only thing that actually necessitates recompilation is configuration updates that affect the #defines for the given endpoint; proxy redirection and PolicyMap updates are one in the same at this point, and we no longer use #defines in the endpoint lxc_config.h for the CIDR-based policy. Is this a correct statement?

**Review by aanm — APPROVED** [review]

**Comment by rlenglet** [comment]
That’s not correct. We still compile in the set of CIDR prefix lengths. So a change in policy may still result in a BPF regeneration.

**Comment by rlenglet** [comment]
> This is still a TODO, but should we ever do this? Wouldn't this require bootstrapping the proxy and having the proxy configured to accept traffic on that port?

You’re right. I’ll remove the TODO.
I still need to fix the CLI to dump the proxy ports from the policy map.

**Comment by ianvernon** [comment]
> That’s not correct. We still compile in the set of CIDR prefix lengths. So a change in policy may still result in a BPF regeneration.

OK, thanks for clarifying. I was being hopeful :)

**Comment by tgraf** [comment]
> That’s not correct. We still compile in the set of CIDR prefix lengths. So a change in policy may still result in a BPF regeneration.

But that should only be the case on older kernels if LPM is not available. Ideally we never have to recompile anymore on policy change

**Comment by ianvernon** [comment]
>  Ideally we never have to recompile anymore on policy change

:D :D :D - this is awesome! I wonder how much this will improve the scalability of policy updates? Also a good motivation for using newer kernels, amongst the many that already exist ;)

**Comment by rlenglet** [comment]
> But that should only be the case on older kernels if LPM is not available. Ideally we never have to recompile anymore on policy change

True, but we generate the `IPCACHE6_PREFIXES` and `IPCACHE4_PREFIXES` macros unconditionally right now.

**Comment by ianvernon** [comment]
> True, but we generate the IPCACHE6_PREFIXES and IPCACHE4_PREFIXES macros unconditionally right now.

We could only do this based on the kernel version, then? I.e., we set some option that we can check at compilation time to not generate them for the headerfile.

**Comment by rlenglet** [comment]
I guess we can just check the map type of the `IPCache` map in `pkg.maps.ipcache.ipcache.go`. If `ipcache.IPCache.MapType == BPF_MAP_TYPE_LPM_TRIE`, then we can skip defining those macros.

I tested in the dev Vagrant VM (4.9.17 kernel, not supporting LPM) that this is a reliable condition.
Will update the PR with a commit doing that.

File: cilium/cmd/helpers.go (hunk 1 above)
Comment (rlenglet) [reply cilium/cmd/helpers.go]: Removed.

**Comment by rlenglet** [comment]
test-me-please

**Comment by rlenglet** [comment]
> consider using a struct for the value of desiredMapState for readability. It's not clear that the map value is the redirect port at first glance.

Done.

**Review by joestringer — APPROVED** [review]
This is awesome. Only thing that made me nervous was the shifting of the `ProxyWaitGroup` / `syncPolicyMap` logic in `pkg/endpoint/bpf.go`.

Probably `bpf/lxc_config.h` needs an update with this change, too.

**Comment by rlenglet** [comment]
test-me-please

**Comment by rlenglet** [comment]
> Probably bpf/lxc_config.h needs an update with this change, too.

I don't think so. That file already doesn't define `CFG_L3L4_INGRESS` nor `CFG_L3L4_EGRESS`.
And we still need to generate `IPCACHE4_PREFIXES` / `IPCACHE6_PREFIXES` on kernel versions <4.11.

**Comment by rlenglet** [comment]
test-me-please

**Comment by rlenglet** [comment]
CI build failure:
https://jenkins.cilium.io/job/Cilium-PR-Ginkgo-Tests-Validated/4709/execution/node/90/log/
```
22:24:55 K8sValidatedDemosTest
22:24:55 /home/jenkins/workspace/Cilium-PR-Ginkgo-Tests-Validated@2/src/github.com/cilium/cilium/test/ginkgo-ext/scopes.go:344
22:24:55   Tests Star Wars Demo [It]
22:24:55   /home/jenkins/workspace/Cilium-PR-Ginkgo-Tests-Validated@2/src/github.com/cilium/cilium/test/ginkgo-ext/scopes.go:376
22:24:55 
22:24:55   unable to access deathstar.default.svc.cluster.local/v1/exhaust-port when policy allows it; 000
22:24:55   Expected
22:24:55       <string>: 000
22:24:55   to contain substring
22:24:55       <string>: 503
22:24:55 
22:24:55   /home/jenkins/workspace/Cilium-PR-Ginkgo-Tests-Validated@2/src/github.com/cilium/cilium/test/k8sT/demos.go:159
```
curl exited with a connection timeout (1 second):
```
STEP: Showing how alliance can access "deathstar.default.svc.cluster.local/v1/exhaust-port" with force header in API request to attack the deathstar
time="2018-07-05T22:23:53Z" level=debug msg="running command: kubectl exec -n default xwing-2768760694-0wj8z -- curl -s --output /dev/stderr -w '%{http_code}' --connect-timeout 1 -X PUT -H 'X-Has-Force: True' http://deathstar.default.svc.cluster.local/v1/exhaust-port"
cmd: "kubectl exec -n default xwing-2768760694-0wj8z -- curl -s --output /dev/stderr -w '%!{(MISSING)http_code}' --connect-timeout 1 -X PUT -H 'X-Has-Force: True' http://deathstar.default.svc.cluster.local/v1/exhaust-port" exitCode: 28 stdout:
000
STEP: Expecting 503 to be returned when using force header to attack the deathstar
=== Test Finished at 2018-07-05T22:23:53Z====
```
Note that the previous test case just before, shows that the Envoy proxy correctly denies a request between the exact same two pods:
```
time="2018-07-05T22:23:52Z" level=info msg="Waiting for cilium endpoints to be ready" invalid=0 pod=cilium-8x3q8 testName=K8sValidatedDemosTest total=6
STEP: Showing how alliance cannot access "deathstar.default.svc.cluster.local/v1/exhaust-port" without force header in API request after importing L7 Policy
time="2018-07-05T22:23:52Z" level=debug msg="running command: kubectl exec -n default xwing-2768760694-0wj8z -- curl -s --output /dev/stderr -w '%{http_code}' --connect-timeout 1 -X PUT http://deathstar.default.svc.cluster.local/v1/exhaust-port"
cmd: "kubectl exec -n default xwing-2768760694-0wj8z -- curl -s --output /dev/stderr -w '%!{(MISSING)http_code}' --connect-timeout 1 -X PUT http://deathstar.default.svc.cluster.local/v1/exhaust-port" exitCode: 0 stdout:
403
stderr:
Access denied
```
So there is no problem with policy enforcement or proxy redirects.

This may be a flake due to the low connection timeout (1 second).
There seem to be issues with kube-dns:
```
cmd: kubectl get pods -o wide --all-namespaces
NAMESPACE     NAME                               READY     STATUS        RESTARTS   AGE       IP              NODE
...
kube-system   kube-dns-2552250787-k2twp          3/3       Running       3          42m       10.10.0.163     k8s1
...
```
Since this test relies on DNS, if kube-dns restarted at the same time as that test case, it could cause the curl command to exceed its 1-second timeout.

In any case, since the immediate previous test case demonstrates correct proxying via Envoy, we can conclude that this test failure is not related to this PR.

**Comment by rlenglet** [comment]
test-me-please

**Comment by rlenglet** [comment]
Another CI build failure:
https://jenkins.cilium.io/job/Cilium-PR-Ginkgo-Tests-Validated/4717/execution/node/90/log/
```
06:10:42 K8sValidatedKafkaPolicyTest
06:10:42 /home/jenkins/workspace/Cilium-PR-Ginkgo-Tests-Validated/src/github.com/cilium/cilium/test/ginkgo-ext/scopes.go:344
06:10:42   Kafka Policy Tests
06:10:42   /home/jenkins/workspace/Cilium-PR-Ginkgo-Tests-Validated/src/github.com/cilium/cilium/test/ginkgo-ext/scopes.go:344
06:10:42     KafkaPolicies [It]
06:10:42     /home/jenkins/workspace/Cilium-PR-Ginkgo-Tests-Validated/src/github.com/cilium/cilium/test/ginkgo-ext/scopes.go:376
06:10:42 
06:10:42     Expected
06:10:42         <string>: ----------------------------------------------------------------
06:10:42         From: [] => To: [] Ports: [80/ANY]
06:10:42           Policy enforcement is disabled because no rules in the policy repository match any endpoint selector from the provided destination sets of labels.
06:10:42         Label verdict: allowed
06:10:42         
06:10:42         Final verdict: ALLOWED
06:10:42         
06:10:42     to contain substring
06:10:42         <string>: Final verdict: DENIED
06:10:42 
06:10:42     /home/jenkins/workspace/Cilium-PR-Ginkgo-Tests-Validated/src/github.com/cilium/cilium/test/k8sT/KafkaPolicies.go:212
```

Tracing doesn't work for any request, because no imported policy rules matches any pod labels passed by the CLI in the tracing request:
```
STEP: Validating Policy trace
time="2018-07-06T06:07:12Z" level=debug msg="running command: kubectl exec -n kube-system cilium-p95ww -- cilium policy trace --src-k8s-pod default:empire-hq-2917507631-4664d --dst-k8s-pod default:kafka-broker-2873197211-xnqlg --dport 9092"
cmd: "kubectl exec -n kube-system cilium-p95ww -- cilium policy trace --src-k8s-pod default:empire-hq-2917507631-4664d --dst-k8s-pod default:kafka-broker-2873197211-xnqlg --dport 9092" exitCode: 0 stdout:
----------------------------------------------------------------
From: [] => To: [] Ports: [9092/ANY]
  Policy enforcement is disabled because no rules in the policy repository match any endpoint selector from the provided destination sets of labels.
Label verdict: allowed

Final verdict: ALLOWED
...
time="2018-07-06T06:07:13Z" level=debug msg="running command: kubectl exec -n kube-system cilium-p95ww -- cilium policy trace --src-k8s-pod default:empire-backup-2986355077-ffthw --dst-k8s-pod default:kafka-broker-2873197211-xnqlg --dport 9092"
cmd: "kubectl exec -n kube-system cilium-p95ww -- cilium policy trace --src-k8s-pod default:empire-backup-2986355077-ffthw --dst-k8s-pod default:kafka-broker-2873197211-xnqlg --dport 9092" exitCode: 0 stdout:
----------------------------------------------------------------
From: [] => To: [] Ports: [9092/ANY]
  Policy enforcement is disabled because no rules in the policy repository match any endpoint selector from the provided destination sets of labels.
Label verdict: allowed

Final verdict: ALLOWED
...
time="2018-07-06T06:07:14Z" level=debug msg="running command: kubectl exec -n kube-system cilium-p95ww -- cilium policy trace --src-k8s-pod default:empire-hq-2917507631-4664d --dst-k8s-pod default:kafka-broker-2873197211-xnqlg --dport 80"
cmd: "kubectl exec -n kube-system cilium-p95ww -- cilium policy trace --src-k8s-pod default:empire-hq-2917507631-4664d --dst-k8s-pod default:kafka-broker-2873197211-xnqlg --dport 80" exitCode: 0 stdout:
----------------------------------------------------------------
From: [] => To: [] Ports: [80/ANY]
  Policy enforcement is disabled because no rules in the policy repository match any endpoint selector from the provided destination sets of labels.
Label verdict: allowed

Final verdict: ALLOWED
...
=== Test Finished at 2018-07-06T06:07:15Z====
```

**NOTE: This test runs explicitly the `cilium policy trace` commands on k8s2 / `cilium-p95ww`, i.e. NOT on the node where all the relevant pods are running (k8s1 / `cilium-vt0hj`).**

```
cmd: kubectl get pods -o wide --all-namespaces
NAMESPACE     NAME                                   READY     STATUS        RESTARTS   AGE       IP              NODE
...
default       empire-backup-2986355077-ffthw         1/1       Running       0          7m        10.10.0.158     k8s1
default       empire-hq-2917507631-4664d             1/1       Running       0          7m        10.10.0.129     k8s1
default       kafka-broker-2873197211-xnqlg          1/1       Running       0          7m        10.10.0.205     k8s1
...
```

The Cilium pod running on `k8s1` is `cilium-vt0hj`. Cilium lists the Kafka broker as managed and ready:
```
time="2018-07-06T06:07:20Z" level=debug msg="running command: kubectl exec -n kube-system cilium-vt0hj -- cilium endpoint list"
cmd: kubectl exec -n kube-system cilium-vt0hj -- cilium endpoint list
ENDPOINT   POLICY (ingress)   POLICY (egress)   IDENTITY   LABELS (source:key[=value])                       IPv6                 IPv4          STATUS
           ENFORCEMENT        ENFORCEMENT
...
32456      Enabled            Disabled          49341      k8s:app=kafka                                     f00d::a0a:0:0:7ec8   10.10.0.205   ready
                                                           k8s:io.cilium.k8s.policy.serviceaccount=default
                                                           k8s:io.kubernetes.pod.namespace=default
                                                           k8s:zgroup=kafkaTestApp
...
```

Before and after the failure, Cilium's policy does contain rules that match on `app=kafka`:
```
$ grep 'app.*kafka' cilium-vt0hj_policy_get.txt 
        "any:app": "kafka",
              "any:app": "kafka",
        "any:app": "kafka",
              "any:app": "kafka",
        "any:app": "kafka",
        "any:app": "kafka",
```
So there is no problem with policy import.

One odd thing is that the Cilium agents on both nodes don't seem to be in sync. They have different policy revision numbers (k8s2 is at revision 94, k8s1 is at revision 90):
```
time="2018-07-06T06:06:16Z" level=info msg="CiliumPolicyAction: Wait for endpoints to sync on pod 'cilium-p95ww'" testName=K8sValidatedKafkaPolicyTest
time="2018-07-06T06:06:16Z" level=debug msg="running command: kubectl exec -n kube-system cilium-p95ww -- cilium policy wait 90"
cmd: "kubectl exec -n kube-system cilium-p95ww -- cilium policy wait 90" exitCode: 0 stdout:

time="2018-07-06T06:06:20Z" level=info msg="CiliumPolicyAction: revision 90 in pod 'cilium-p95ww' is ready" testName=K8sValidatedKafkaPolicyTest
time="2018-07-06T06:06:20Z" level=info msg="CiliumPolicyAction: Wait for endpoints to sync on pod 'cilium-vt0hj'" testName=K8sValidatedKafkaPolicyTest
time="2018-07-06T06:06:20Z" level=debug msg="running command: kubectl exec -n kube-system cilium-vt0hj -- cilium policy wait 89"
cmd: "kubectl exec -n kube-system cilium-vt0hj -- cilium policy wait 89" exitCode: 0 stdout:
^MWaiting for endpoints to run policy revision 89: 4/5              ^MWaiting for endpoints to run policy revision 89: 4/5              ^MWaiting for endpoints to run policy revision 89: 4/5              ^MWaiting for endpoints to run policy revision 89: 4/5              ^MWaiting for endpoints to run policy revision 89: 0/5              ^MWaiting for endpoints to run policy revision 89: 0/5

time="2018-07-06T06:07:03Z" level=info msg="CiliumPolicyAction: revision 89 in pod 'cilium-vt0hj' is ready" testName=K8sValidatedKafkaPolicyTest
time="2018-07-06T06:07:04Z" level=debug msg="running command: kubectl exec -n kube-system cilium-p95ww -- cilium endpoint list -o jsonpath='{range [*]}{@.id}{\"_\"}{@.status.networking.addressing[0].ipv4}{\"=\"}{@.status.policy.spec.policy-revision}{\"\\n\"}{end}'"
cmd: "kubectl exec -n kube-system cilium-p95ww -- cilium endpoint list -o jsonpath='{range [*]}{@.id}{\"_\"}{@.status.networking.addressing[0].ipv4}{\"=\"}{@.status.policy.spec.policy-revision}{\"\\n\"}{end}'" exitCode: 0 stdout:
2412_10.10.1.200=94
8904_10.10.1.12=94
18104_10.10.1.82=94
27090_10.10.1.63=94
38494_10.10.1.163=94
62270_10.10.1.216=94

time="2018-07-06T06:07:06Z" level=debug msg="running command: kubectl exec -n kube-system cilium-vt0hj -- cilium endpoint list -o jsonpath='{range [*]}{@.id}{\"_\"}{@.status.networking.addressing[0].ipv4}{\"=\"}{@.status.policy.spec.policy-revision}{\"\\n\"}{end}'"
cmd: "kubectl exec -n kube-system cilium-vt0hj -- cilium endpoint list -o jsonpath='{range [*]}{@.id}{\"_\"}{@.status.networking.addressing[0].ipv4}{\"=\"}{@.status.policy.spec.policy-revision}{\"\\n\"}{end}'" exitCode: 0 stdout:
30902_10.10.0.158=90
32456_10.10.0.205=90
51950_10.10.0.96=90
59450_10.10.0.129=90
61739_10.10.0.190=90
```
But that seems inconsequential. `cilium-vt0hj` does successfully import the policy with rules that match `app=kafka`.

So this issue may be due to the CLI passing wrong labels to the tracing API call.
It's impossible to know what labels are passed to the agent in the trace request, because the logs are garbage, and the parameters are never logged:
```
$ grep 'GET /policy/resolve request' pod-kube-system-cilium-p95ww-cilium-agent.log 
2018-07-06T06:07:13.410412845Z level=debug msg="GET /policy/resolve request" params="{HTTPRequest:0xc423138500 TraceSelector:0xc423313380}"
2018-07-06T06:07:14.263388291Z level=debug msg="GET /policy/resolve request" params="{HTTPRequest:0xc423138e00 TraceSelector:0xc4232c9740}"
2018-07-06T06:07:15.08570593Z level=debug msg="GET /policy/resolve request" params="{HTTPRequest:0xc421ee8800 TraceSelector:0xc423288a60}"
```

When receiving `--src-k8s-pod` or `--dst-k8s-pod` flags, the CLI does a `GET /identity/{id}` request to the agent to get the labels for the pod. Those labels are the ones that are passed in the tracing request. However, neither the client nor the agent logs anything about those `GET /identity/{id}` requests, so it's impossible to know which labels are then passed for tracing.

After the test failure, the pods seem to have the right `cilium.io/identity` annotations, e.g.:
```
Name:           kafka-broker-2873197211-xnqlg
Namespace:      default
Node:           k8s1/192.168.36.11
Start Time:     Fri, 06 Jul 2018 05:59:59 +0000
Labels:         app=kafka
                pod-template-hash=2873197211
                zgroup=kafkaTestApp
Annotations:    cilium.io/identity=49341
...
```
```
$ grep cilium.io/identity pods.txt 
                    "cilium.io/identity": "35914",
                    "cilium.io/identity": "11708",
                    "cilium.io/identity": "19115",
                    "cilium.io/identity": "9318",
                    "cilium.io/identity": "49341",
                    "cilium.io/identity": "47838",
                    "cilium.io/identity": "59029",
                    "cilium.io/identity": "22110",
```

There is no issue with the timing of the setting of the `cilium.io/identity`, e.g. for the Kafka broker:
```
$ grep 'Successfully annotated pod with cilium.io/identity' pod-kube-system-cilium-vt0hj-cilium-agent.log | grep endpointID=32456 | head -n 1
2018-07-06T06:00:01.777625334Z level=debug msg="Successfully annotated pod with cilium.io/identity=49341" endpointID=32456 k8sIdentityAnnotation=cilium.io/identity k8sNamespace=default k8sPodName=kafka-broker-2873197211-xnqlg retryUUID=d23406c9-80e1-11e8-a672-080027ec2d6c subsys=k8s
```
That is, the annotation is set for the first time, and always with the correct identity, at 06:00:01, whereas the first tracing failure happens at 06:07:14.

Anyway, if the identity was unknown to `cilium-p95ww`, the identity get request would have failed, and the command would have failed as well.


The real problem here seems to be that this test does a policy trace on a different node from all the pods that being tested, which results in querying a node which may still be out of sync re: policy or identities. I'll submit a new commit in this PR to do the tracing on the same node (k8s1).

**Comment by rlenglet** [comment]
test-me-please

**Comment by rlenglet** [comment]
test-me-please

**Comment by rlenglet** [comment]
test-me-please

**Review by ianvernon — APPROVED** [review]

**Pull request merged**